### PR TITLE
Add noDirectMutationState rule and tests

### PR DESCRIPTION
**Pull request by stefanuros**
#341 
Implemented [react/no-direct-mutation-state](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-direct-mutation-state.md). 

If anyone has any ideas for scenarios that should be tested, please let me know. This was my first contribution so any advice/suggestions about the code are welcome.

### CODE REVIEW

**Comment by stefanuros** [comment]
Currently unable to continue as this error is coming up when I run `scripts/dev-rome lint`. I am working to fix this and once it is, I will make the PR ready for review.

```
Unhandled CLI query error
Error: raw and highlighted line count mismatch 841 !== 966
  at ___R$project$rome$$romejs$cli$diagnostics$utils_ts$toLines (project-rome/@romejs/cli-diagnostics/utils.ts:80:8)
  at ___R$project$rome$$romejs$cli$diagnostics$DiagnosticsPrinter_ts$default.addFileSource (project-rome/@romejs/cli-diagnostics/DiagnosticsPrinter.ts:241:12)
  at ___R$project$rome$$romejs$cli$diagnostics$DiagnosticsPrinter_ts$default.fetchFileSources (project-rome/@romejs/cli-diagnostics/DiagnosticsPrinter.ts:344:9)
  at ___R$project$rome$$romejs$cli$diagnostics$DiagnosticsPrinter_ts$default.print (project-rome/@romejs/cli-diagnostics/DiagnosticsPrinter.ts:351:7)
  at ___R$project$rome$$romejs$core$master$Master_ts$default.handleRequestError (project-rome/@romejs/core/master/Master.ts:961:12)
  at ___R$project$rome$$romejs$core$master$Master_ts$default.dispatchRequest (project-rome/@romejs/core/master/Master.ts:888:51)
  at global.runMicrotasks
  at process.processTicksAndRejections (internal/process/task_queues.js:97:4) generated source location
```

**Comment by sebmck** [comment]
Can you rebase and try again? #583 was merged that should resolve that error however it will uncover another that might be more actionable.

**Comment by stefanuros** [comment]
I tried that, unfortunately no change. Its something to do with the noDirectMutationState.ts file. If I remove most of the file, it stops giving me that error.

**Comment by stefanuros** [comment]
@sebmck I think it has something to do with `doesNodeMatchPattern(node.left, "this.state")` on line 27. When I removed that statement, running `scripts/dev-rome lint` works fine. Perhaps I am misunderstanding about how to use `doesNodeMatchPattern` or what `JSAssignmentExpression` contains?

**Comment by macovedj** [comment]
Did some looking, had some thoughts.  fwiw if I just pull your branch I don't get the error.
Also I think your implementation takes issue with general mutation of state in a class, not necessarily in a react component.  When you run the lint rule, it thinks that rome itself is in violation, even though rome isn't written in react.

**Comment by stefanuros** [comment]
The new commit makes sure the node being considered is a react component. React components are defined by a class extending `React.Component`, a class extending `Component`, or being withing a function call where the identifier is `createReactClass`. This new change has removed the raw vs highlighted issue I was having (still unsure what was causing it). If I am missing any tests I should be checking, or any different ways of defining react components, please let me know.

**Comment by sebmck** [comment]
Hey @stefanuros! I haven't forgotten about this PR. I just haven't had time to review it. I'll try to find some time later today.

File: packages/@romejs/diagnostics/descriptions/lint.ts [hunk 1]
```diff
@@ -457,6 +457,17 @@ export const lint = createDiagnosticsCategory({
...
 		],
 	},
+	REACT_NO_DIRECT_MUTATION_STATE: {
+		category: "lint/react/noDirectMutationState",
+		message: "Avoid mutating <emphasis>this.state</emphasis> directly.",
+		advice: [
+			{
+				type: "log",
+				category: "info",
+				text: "Calling <emphasis>setState()</emphasis> after mutating <emphasis>this.state</emphasis> directly may replace the mutation you made.",
```
Comment (VictorHom) [inline packages/@romejs/diagnostics/descriptions/lint.ts:467]: I think adding an additional "The only place to set this.setState directly is in constructor of a react class component."

File: packages/@romejs/compiler/lint/rules/react/noDirectMutationState.test.ts [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+
+test(
+	"Prevent direct mutation of this.state",
+	async (t) => {
+		await testLint(
+			t,
+			{
+				invalid: [
+					`
+            var Hello = createReactClass({
```
Comment (VictorHom) [inline packages/@romejs/compiler/lint/rules/react/noDirectMutationState.test.ts:12]: awesome on the extensive test cases! from a decision in previous discussions (https://github.com/romejs/rome/issues/341#issuecomment-632264987 as one place), you don't need to account for the `createReactClass` case. Sorry for the noise

Comment (stefanuros) [reply packages/@romejs/compiler/lint/rules/react/noDirectMutationState.test.ts]: Whoops I missed that. Ill take those checks out. Ive changed the createReactClass tests as well and stripped a couple other repetitive ones.

**Comment by stefanuros** [comment]
@VictorHom  Awesome, thanks for your help!

**Pull request merged**